src/Symfony/Component/Messenger/Transport/Doctrine/DoctrineReceiver.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Doctrine;
+
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\Serializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class DoctrineReceiver implements ReceiverInterface
+{
+    private $connection;
+    private $serializer;
+    private $shouldStop = false;
+
+    public function __construct(Connection $connection, SerializerInterface $serializer = null)
+    {
+        $this->connection = $connection;
+        $this->serializer = $serializer ?? Serializer::create();
+    }
+
+    public function receive(callable $handler): void
+    {
+        while (!$this->shouldStop) {
+            $doctrineEnvelope = $this->connection->get();
+
+            if (null === $doctrineEnvelope) {
+                $handler(null);
+
+                usleep(200000);
+                $this->pcntlDispatch();
+
+                continue;
+            }
+
+            try {
+                $handler($this->serializer->decode(array(
+                    'body' => $doctrineEnvelope['body'],
+                    'headers' => $doctrineEnvelope['headers'],
+                )));
+
+                $this->connection->ack($doctrineEnvelope['id']);
+            } catch (\Throwable $e) {
+                $this->connection->nack($doctrineEnvelope['id']);
+
+                throw $e;
+            } finally {
+                $this->pcntlDispatch();
+            }
+        }
+    }
+
+    public function stop(): void
+    {
+        $this->shouldStop = true;
+    }
+
+    private function pcntlDispatch()
+    {
+        if (\function_exists('pcntl_signal_dispatch')) {
```

Comment:
This is a problem of current implementations as well, but does it make much sense to check for pcntl every loop, since this can't change during the run of the program?

Reply:
I've added a new property on the class that is initialized on the construct but I'me not sure it's good for readability 🤔

Reply:
tbh I would use in this case normal variable and inline the function

Reply:
Hum I don't get what is your idea here 🤔. Just get the result of `function_exists` into a variable in the `receive` method instead of create a `pcntlDispatch` method ?

Reply:
Correct. Inlining function is this

```diff
    } finally {
-        $this->pcntlDispatch();
+        if ($pcntlAvailable) {
+           pcntl_signal_dispatch();
+        }
    }
}

- private function pcntlDispatch()
- {
-     if ($this->pcntlAvailable) {
-         pcntl_signal_dispatch();
-     }
- }
```

I'm aware it's used two times, but it doesn't provide enough value.

Reply:
Then I go do as it was at the begining like on AmqpReceiver : https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php

This way the two receiver are similar